packages/react-aria-components/src/Toolbar.tsx

```diff
@@ -0,0 +1,176 @@
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+function useToolbar(props: ToolbarProps & {isInsideAToolbar: boolean}, ref: ForwardedRef<HTMLDivElement>) {
+  const {
+    'aria-label': ariaLabel,
+    'aria-labelledby': ariaLabelledBy,
+    orientation = 'horizontal',
+    isInsideAToolbar
+  } = props;
+  const {direction} = useLocale();
+  const shouldReverse = direction === 'rtl' && orientation === 'horizontal';
+  const focusManager = useFocusManager();
+
+  const onKeyDown: KeyboardEventHandler = (e) => {
+    if (
+      (orientation === 'horizontal' && e.key === 'ArrowRight')
+      || (orientation === 'vertical' && e.key === 'ArrowDown')) {
+      if (shouldReverse) {
+        focusManager.focusPrevious();
+      } else {
+        focusManager.focusNext();
+      }
+    } else if (
+      (orientation === 'horizontal' && e.key === 'ArrowLeft')
+      || (orientation === 'vertical' && e.key === 'ArrowUp')) {
+      if (shouldReverse) {
+        focusManager.focusNext();
+      } else {
+        focusManager.focusPrevious();
+      }
+    } else if (e.key === 'Tab') {
+      // When the tab key is pressed, we want to move focus
+      // out of the entire toolbar. To do this, move focus
+      // to the first or last focusable child, and let the
+      // browser handle the Tab key as usual from there.
+      e.stopPropagation();
+      lastFocused.current = document.activeElement as HTMLElement;
+      if (e.shiftKey) {
+        focusManager.focusFirst();
+      } else {
+        focusManager.focusLast();
+      }
+      return;
+    } else {
+      // if we didn't handle anything, return early so we don't preventDefault
+      return;
+    }
+
+    // Prevent arrow keys from being handled by nested action groups.
+    e.stopPropagation();
+    e.preventDefault();
+  };
+
+  // Record the last focused child when focus moves out of the toolbar.
+  const lastFocused = useRef<HTMLElement | null>(null);
+  const onBlur = (e) => {
+    if (!e.currentTarget.contains(e.relatedTarget) && !lastFocused.current) {
+      lastFocused.current = e.target;
+    }
+  };
+
+  // Restore focus to the last focused child when focus returns into the toolbar.
+  // If the element was removed, do nothing, either the first item in the first group,
+  // or the last item in the last group will be focused, depending on direction.
+  const onFocus = (e) => {
+    if (lastFocused.current && !e.currentTarget.contains(e.relatedTarget)) {
+      lastFocused.current?.focus();
+      lastFocused.current = null;
+    }
+  };
+
+  return {
+    toolbarProps: {
+      role: !isInsideAToolbar ? 'toolbar' : undefined,
+      'aria-orientation': orientation,
+      'aria-label': ariaLabel,
+      'aria-labelledby': ariaLabel == null ? ariaLabelledBy : undefined,
+      onKeyDownCapture: !isInsideAToolbar ? onKeyDown : undefined,
+      onFocusCapture: !isInsideAToolbar ? onFocus : undefined,
+      onBlurCapture: !isInsideAToolbar ? onBlur : undefined
```

Comment:
now that i have a context and turn off the keyboard handlers for nested toolbars and actiongroups, I could theoretically turn off capture here

this would present a problem if someone tried to use an ActionGroup from before this release

It would allow us to put a Picker/Select inside of the group which could be opened via the down arrow. In orientation vertical though, this might be a problem. How should a Picker/Select be opened when inside a toolbar? Asking accessibility

Reply:
right I think we would want to use capture anyway so that non-ActionGroups can be supported in children.